training/CHAPEL_TUTORIAL.md

```diff
@@ -0,0 +1,861 @@
+so the end bound is determined by the size of the iterables.
+Since in this case the other iterables are length 5, `1..` is equivalent to `1..5`.
+
+<a id="ex15"></a>
+#### Example 15: Zippered Iteration in Arkouda
+Now let's look at an example based on [`getLengths`](https://github.com/Bears-R-Us/arkouda/blob/b86adeb843275b0b86553534ede632acef4d15e2/src/SegmentedString.chpl#L388-L406) in `SegmentedString.chpl`.
+
+First we need some context. An Arkouda `Strings` (aka SegString) is an array of variable length strings.
+```python
+>>> s = ak.array([f"{'s'*(3-i)}{i}" for i in range(3)])
+>>> s
+array(['sss0', 'ss1', 's2'])
+>>> type(s)
+arkouda.strings.Strings
+```
+
+On the server, a SegString is made up of two distributed array components:
+- values: `uint(8)`
+  - the flattened array of bytes in all the strings (null byte delimited)
+- offsets: `int`
+  - the start index of each individual string in the `values`
+
+For the sake of simplicity, we'll treat values as a string array. So for our `s` above, these components look something like
+```
+values = ['s', 's', 's', '0', '\x00', 's', 's', '1', '\x00', 's', '2', '\x00']
+offsets = [0, 5, 9]
+```
+For `getLengths`, we want to calculate the length of each individual string including the null terminator.
+
+```chapel
+const values: [0..#12] string = ['s', 's', 's', '0', '\x00', 's', 's', '1', '\x00', 's', '2', '\x00'];
+const offsets = [0, 5, 9];
+const size = 3;
+
+/* Return lengths of all strings, including null terminator. */
+proc getLengths() {
+  // initialize lengths with the same domain as offsets
+  var lengths: [offsets.domain] int;
+  if size == 0 {
+    // if the segstring is empty, the lengths are also empty
+    return lengths;
+  }
+  // save off the last index of offsets
+  const high = offsets.domain.high;
+  forall (i, o, l) in zip(offsets.domain, offsets, lengths) {
+    if i == high {
+      // the last string
+      // len = values.size - start position of this string
+      l = values.size - o;
+    }
+    else {
+      // not the last string
+      // len = start poistion of next string - start position of this string
+      l = offsets[i+1] - o;
+    }
+  }
+  return lengths;
+}
+
+writeln(getLengths());
+```
+```console
+5 4 3
+```
+Nice! We'll use `getLengths`, `values`, and `offsets` in the next section as well.
+
+<a id="aggregation"></a>
+### Aggregation
+
+[Aggregation](https://chapel-lang.org/docs/modules/packages/CopyAggregation.html#copyaggregation) is used to copy
+local values into a remote array (`DstAggregator`) or copy values from a remote array into local variables (`SrcAggregator`).
+This provides a significant speed up when doing batch assignments.
+
+Syntax:
+```chapel
+// without aggregation
+forall (i, v) in zip(inds, vals) {
+  remoteArr[i] = v;
+}
+
+// with aggregation
+use CopyAggregation;
+forall (i, v) in zip(inds, vals) with (var agg = new DstAggregator(int)) {
```

Comment:
This starts using `with (var ...)` somewhat out of the blue.  It could be useful to introduce this earlier, when the forall loops were first introduced, as introducing a task-local variable.  That'd make it a nice companion to `ref` and/or `reduce` intents, each of which are also introducing task-private variables of one form or another.  Alternatively, could try to describe it briefly here.

Reply:
good point!